### DISCUSSION TITLE
Supporting AI/NLP & text mining based workflows

### DISCUSSION THREAD

**Discussion by dmrd** [athensresearch/athens#1337] [Feature Requests]
After some initial discussion @ the community calls, I want to start a discussion here around what it would look like to integrate natural language tools, ranging from semantic search to GPT-style language models (LMs) into Athens. I mainly talk about language models below, but there's no reason to limit exploration & design to this class of models:  there is a lot of room for much simpler solutions (many without any learning at all!) to support Athens users in writing and creating.

The core point that I believe makes this useful is _**discrimination is easier than generation**_ [(obligatory reference to that one ira glass quote)](https://www.goodreads.com/quotes/309485-nobody-tells-this-to-people-who-are-beginners-i-wish)
   
A couple examples of what text completion can look like in writing & creative endeavors:
    - [ Sudowrite](https://www.sudowrite.com/) applies this to creative writing: when you encounter writers block, the model proposed possible completions - editing/choosing is often easier than generating!
    -  ldea crossover & brainstorming: Given a set of blocks, what ideas are similar? Different? This is illustrated on words (rather than whole sentences) in learned word embeddings, where word analogies can be carried out in vector space, such as [king - man + woman = queen](https://blog.acolyer.org/2016/04/21/the-amazing-power-of-word-vectors/) 
    -  David Bieber has a nice post on experimenting with GPT3, which naturally begins with using GPT3 to generate more ideas on what to use it for: [https://davidbieber.com/snippets/2020-07-22-writing-with-gpt3/](https://davidbieber.com/snippets/2020-07-22-writing-with-gpt3/)

In terms of implementation, one can directly integrate models using web APIs, such as GPT3 through OpenAI and many smaller models through [Huggingface](https://huggingface.co/). Models somewhere between GPT2 & GPT3 are starting to be released into the public domain (see: [GPT-J](https://www.reddit.com/r/MachineLearning/comments/nvkowg/p_gptj_6b_jaxbased_transformer_lm/))

For many use cases, applying ideas of search engines (semantic embeddings, knowledge mining) will seem just as magical: imagine a sidebar in Athens presenting you with the most relevant snippets across your DB as you write, even if they only overlap in semantics (meaning) not syntax (like the words you used). See the Semantica and MemNav writeup on the Psionica site (linked below).

I recommend the writing and experiments that @paulbricman has been doing in his Psionica project around avenues to integrate simple natural language tools to augment various aspects of writing and using a personal knowledge base.
    - [ Dual | Psionica](https://psionica.org/docs/workshop/dual/): Formulates a language model agent as a knowledge base assistant, allowing you to teach new skills ("turn this note into flashcards")
    - [ K-Probes | Psionica](https://psionica.org/docs/lab/k-probes/): simple pre-written prompts can be enough to help generation, no statistical methods required.
    - [ Semantica | Psionica](https://psionica.org/docs/lab/semantica/): uses semantic embeddings (like the word analogy example above) to generate options & ideas.
    - [ MemNav | Psionica](https://psionica.org/docs/lab/memnav/): Apply text mining methods to a personal knowledge base. What if you could ask questions of your DB such as "Who have I talked to about Athens?"

I hope this can spark further discussion - there's no system that takes these sort of assistive tools seriously yet & they will feel truly magical. What workflows do you aspire to in Athens, and what does it take to make them happen?

**Comment by paulbricman** [comment]
Thanks for the tag, @dmrd! Let me know if I can help with brainstorming NLP tools for Athens or with advice on the local NLP tech stack (beyond HuggingFace). All Psionica projects are open source and local-first, so supporting Athens on this path seems like an obvious way to go! One thing I'm personally considering is porting Dual from Obsidian to Athens.
